### PR TITLE
Record.is_record/2 fix that deals with invalid records

### PR DESCRIPTION
**Pull request by eksperimental**

### CODE REVIEW

File: bin/elixir-env [hunk 1]
```diff
@@ -0,0 +1,108 @@
+#!/bin/sh
```
Comment (josevalim) [inline bin/elixir-env]: ?

File: lib/elixir/lib/record.ex [hunk 2]
```diff
@@ -85,14 +85,20 @@ defmodule Record do
...
+            :erlang.is_atom(:erlang.element(1, unquote(data))) and
+            :erlang.element(1, unquote(data)) == unquote(kind)
         end
       false ->
         quote do
           result = unquote(data)
-          is_tuple(result) and tuple_size(result) > 0
-            and :erlang.element(1, result) == unquote(kind)
+          cond do
+            :erlang.is_tuple(result) and :erlang.tuple_size(result) > 0 ->
```
Comment (josevalim) [inline lib/elixir/lib/record.ex]: We don't need to convert it into a cond, do `kind = unquote(kind)` before and check if kind is an atom.

**Comment by josevalim** [comment]
Btw, why do we need to use Erlang functions?

File: bin/elixir-env (hunk 1 above)
Comment (eksperimental) [inline bin/elixir-env]: oops. that's a shell script I'm working on. it shouldn't have shown up here :-P

Comment (josevalim) [inline bin/elixir-env]: Spoiler alert!!!

**Comment by eksperimental** [comment]
Actually, I was going to inquire about that José.
I noticed that in `Kernel`, all guard safe functions were always implemented using Erlang functions. I assumed it was a matter of optimization or that the Elixir version not available at the time of calling the macro.
what's the reason behind?
It's weird though that we are mixing Elixir and Erlang functions.

File: lib/elixir/lib/record.ex (hunk 2 above)
Comment (eksperimental) [inline lib/elixir/lib/record.ex]: great suggestion

**Comment by eksperimental** [comment]
@josevalim changes pushed.
please let me know if you are OK with not mixing erlang and elixir functions. If not I will switch back to :erlang.element/2

File: CHANGELOG.md [hunk 3]
```diff
@@ -88,8 +88,9 @@
   * [Kernel] Raise if `defstruct` is called multiple times
   * [Kernel] Ensure `Module.create/3` respects var/alias hygiene
   * [Kernel] Support non-literal ranges on the right side of `in/2`
-  * [OptionParser] Allow `OptionParser` to parse negative numbers
   * [Macro] Fix `Macro.to_string/1` on a call of a capture argument, for example `&(&1).(:x)`
+  * [OptionParser] Allow `OptionParser` to parse negative numbers
+  * [Record] Fix `Record.is_record/2` when dealing with non-record tuples.
```
Comment (josevalim) [inline CHANGELOG.md]: No trailing dot please!

**Comment by josevalim** [comment]
Just one last comment and we are good to go!

File: CHANGELOG.md (hunk 3 above)
Comment (eksperimental) [inline CHANGELOG.md]: done

**Pull request merged**